src/mods.cpp

```diff
@@ -356,3 +353,78 @@ Json::Value getModstoreUrl(std::string url)
 }
 
 #endif
+
+ModMetadata::ModMetadata(const std::string &mod_name):
+	m_mod_name(mod_name),
+	m_modified(false)
+{
+	m_stringvars.clear();
+}
+
+void ModMetadata::clear()
+{
+	Metadata::clear();
+	m_modified = true;
+}
+
+bool ModMetadata::save(const std::string &root_path)
+{
+	Json::Value json;
+	for (StringMap::const_iterator it = m_stringvars.begin();
+			it != m_stringvars.end(); ++it) {
+		json[it->first] = it->second;
+	}
+
+	if (!fs::PathExists(root_path)) {
+		fs::CreateAllDirs(root_path);
+	} else {
+		if (!fs::IsDir(root_path)) {
+			errorstream << "ModMetadata[" << m_mod_name << "]: Unable to save. '"
+				<< root_path << "' is not a directory." << std::endl;
+			return false;
+		}
+	}
+
+	bool w_ok = fs::safeWriteToFile(root_path + DIR_DELIM + m_mod_name,
+		Json::FastWriter().write(json));
+
+	if (!w_ok) {
+		errorstream << "ModMetadata[" << m_mod_name << "]: failed write file." << std::endl;
+	} else {
+		m_modified = false;
+	}
+	return w_ok;
+}
+
+bool ModMetadata::load(const std::string &root_path)
+{
+	m_stringvars.clear();
+
+	std::ifstream is((root_path + DIR_DELIM + m_mod_name).c_str(), std::ios_base::binary);
+	if (!is.good()) {
+		return false;
+	}
+
+	Json::Reader reader;
+	Json::Value root;
+	if (!reader.parse(is, root)) {
+		errorstream << "ModMetadata[" << m_mod_name << "]: failed read data "
+			"(Json decoding failure)." << std::endl;
+		return false;
+	}
+
+	const Json::Value::Members attr_list = root.getMemberNames();
+	for (Json::Value::Members::const_iterator it = attr_list.begin();
+		it != attr_list.end(); ++it) {
+		Json::Value attr_value = root[*it];
+		m_stringvars[*it] = attr_value.asString();
+	}
+
+	return true;
+}
+
+bool ModMetadata::setString(const std::string &name, const std::string &var)
+{
+	m_modified = Metadata::setString(name, var);
```

Comment:
I'm wondering whether it may be worth making `m_modified` a property of Metadata, and removing the need for `virtual` instead